doc/Neighbor/Neighbor_Refresh_HLD.md

```diff
@@ -0,0 +1,321 @@
+- subscribe to STATE_DB FDB table to manage the mac address table local cache
+- periodically refresh ARP/ND by sending ARP request pkt / NS pkt
+
+
+#### 3.3.2.1 Local Cache
+
+Neighbor Refresh Cache:
+
+- Add neighbor entries to cache when the entry is added to NEIGH_REFRESH_TABLE
+    - All Dynamically learned neighbor entries [ARP, ND (Global, LinkLocal)]
+    - All Static neighbor entries (MAC can be dynamic)
+- Following Entries will not be added to the neighbor cache
+    - Neighbors learned from “eth0” interface
+    - Neighbors learned from BGP/EVPN MAC/IP type-2 route 
+- Remove entries from cache when the entry is deleted from NEIGH_REFRESH_TABLE
+
+```
+v4/v6 Neighbors Cache [map]
+    
+    Key     : IP Address + InterfaceName [Phy/PortChannel/Vlan/Sag]
+
+    Value
+            : MAC Address
+            : IP address family
+            : State  (Reachable/Stale/Failed)
+            : Kernel Flags (For NTF_EXT_LEARNED)
+            : Timestamp (Entry creation/last refresh)
+```
+
+Interface Cache:
+
+- Interface IP address will be used in building the ARP (Sender IP address) and NS (Soure IP) packets
+- Subscribe to redis-db tables
+    - IP address 
+      - CONFIG_DB: INTERFACE, VLAN_INTERFACE, SAG_INTERFACE, PORTCHANNEL_INTERFACE
+
+```
+Interface Cache [Map]
+
+    Key     : Interface name
+    Value   : List of IP address configured
+```
+
+
+FDB Cache:
+
+- FDB Cache is to handle quick refresh/retry of neighbors whose mac address is removed admistratively or due to control protocol clear 
+- It has the list of neighbors(v4/v6) using the same mac on a VLAN interface. When the mac address is deleted, all the neighbors associated with the mac will be added to a fdb refresh queue.
+- Entries in FDB refresh queue will be refreshed faster than the usual neighbor refresh interval and will be removed from the queue after 10 retries.
+
+```
+FDB Cache [Map]
+
+    Key     : Interface + MAC
+    Value   : List of neighbors using this mac
+```
+
+#### 3.3.2.2 Sending ARP/ND Refresh Packets
+
+- For Resolved ARP/ND, Dst MAC will be same as in its NeighCache
+- For Unresolved ARP, Dst MAC will be Broadcast MAC
+- For Unresolved ND, ICMPv6 Echo Request pkt will be sent from application, Kernel will send NS
+- ARP/NS packets will be sent using raw socket 
+  - Separate sockets for ARP and ICMPv6 NS/Echo Request
+- Kernel processing
+  - VLAN tagging & FDB lookup happens based on outgoing interface
+  - On Vlan Interfaces based on the FDB lookup packet will be either flooded or sent to a specific port
+- Soure MAC 
+  - For SAG interfaces, SAG MAC will be used and it will retrived from CONFIG_DB: SAG_GLOBAL
+  - For all other interfaces, Source MAC will be retrived from kernel.
+- Ifname to Index (required for socket send)
+
+#### 3.3.2.3 Refresh Timer
+
+- Traverse the neighbor Cache entries periodically (every 30 secs)
+  - Check refresh timeout has elapsed for every neighbors
+  - If elapsed then send ARP/NS packet
```

Comment:
Is this unicast or broadcast ARP or multicast NS?

Reply:
When the MAC is refreshed in the kernel, there can be a possibility flooding the packets on all members of the VLAN, how are we avoiding this?

Reply:
If the MAC is known then the Neighbor refresh will be unicast, if the MAC is not know then the neighbor refresh will be broadcast ARP /multicast NS.

As long as we have same MAC aging time in both the Kernel and the HW, we should not have this issue.